operators/install_dependencies.py

```diff
@@ -22,33 +21,55 @@ def install_pip():
     except subprocess.CalledProcessError:
+        if not user_site:
+            no_user = os.environ.get("PYTHONNOUSERSITE", None)
+            os.environ["PYTHONNOUSERSITE"] = "1"
         import ensurepip
 
-        ensurepip.bootstrap()
-        os.environ.pop("PIP_REQ_TRACKER", None)
+        try:
+            ensurepip.bootstrap()
+        finally:
+            os.environ.pop("PIP_REQ_TRACKER", None)
+            if not user_site:
+                if no_user:
+                    os.environ["PYTHONNOUSERSITE"] = no_user
+                else:
+                    del os.environ["PYTHONNOUSERSITE"]
 
-def install_and_import_requirements(requirements_txt=None):
+def install_and_import_requirements(requirements_txt=None, user_site=False):
     """
     Installs all modules in the 'requirements.txt' file.
     """
     environ_copy = dict(os.environ)
-    environ_copy["PYTHONNOUSERSITE"] = "1"
-
-    python_devel_tgz_path = absolute_path('python-devel.tgz')
-    response = requests.get(f"https://www.python.org/ftp/python/{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}/Python-{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}.tgz")
-    open(python_devel_tgz_path, 'wb').write(response.content)
-    python_devel_tgz = tarfile.open(python_devel_tgz_path)
+    if not user_site:
+        environ_copy["PYTHONNOUSERSITE"] = "1"
     python_include_dir = sysconfig.get_paths()['include']
-    def members(tf):
-        prefix = f"Python-{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}/Include/"
-        l = len(prefix)
-        for member in tf.getmembers():
-            if member.path.startswith(prefix):
-                member.path = member.path[l:]
-                yield member
-    python_devel_tgz.extractall(path=python_include_dir, members=members(python_devel_tgz))
+    if not os.path.exists(python_include_dir):
+        try:
+            os.makedirs(python_include_dir)
+        finally:
+            pass
+    if os.access(python_include_dir, os.W_OK):
+        print("downloading additional include files")
+        python_devel_tgz_path = absolute_path('python-devel.tgz')
+        response = requests.get(f"https://www.python.org/ftp/python/{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}/Python-{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}.tgz")
+        open(python_devel_tgz_path, 'wb').write(response.content)
+        python_devel_tgz = tarfile.open(python_devel_tgz_path)
+        def members(tf):
+            prefix = f"Python-{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}/Include/"
+            l = len(prefix)
+            for member in tf.getmembers():
+                if member.path.startswith(prefix):
+                    member.path = member.path[l:]
+                    yield member
+        python_devel_tgz.extractall(path=python_include_dir, members=members(python_devel_tgz))
```

Comment:
@carson-katri What exactly is this for? I've noticed no issues during installing and running without this step.

Reply:
It fixed an error I had when initially developing, but I don’t remember exactly why. Perhaps needed when building from source on certain platforms.